WordPress/gutenberg
# Patterns: create patterns package and share the create pattern modal between post editor and site editor

## What?
Shares a create pattern modal between the post and the site editor via a newly created `patterns` package. This new package will take over all management of synced and unsynced user created patterns to allow the eventual deprecation of the `reusable-blocks` package.

## Why?
Currently there is duplicated code for adding patterns in post editor (imported from `reusable-blocks`) and site editor (local modal component), which means any changes need to be applied in two places.

## How?
Adds a CreatePatternModel in a new `patterns` package.

## Testing Instructions

- In the post editor add a block and use the `Create pattern` block menu to create both a synced and an unsynced pattern and make sure both options work as expected
- In the post editor make sure the Detach pattern and Manage patterns options on the block overflow menu work correctly
- In the site editor pattern section use the + at the top of the left navigation to create both a synced and unsynced pattern and make sure both are created as expected
- In the site editor block editor use the `Create pattern` block menu to create both a synced and an unsynced pattern and make sure both options work as expected and also make sure the Detach pattern and Manage patterns options on the block overflow menu work correctly

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/d5d7b369-5369-48e3-801c-95d3bdd615f3
